### DISCUSSION TITLE
Managing function declarations in typeclass instances (here Functor)

### DISCUSSION THREAD

**Discussion by ornamentist** [coalton-lang/coalton#674] [Help!] [answered]
I'm probably missing something obvious here, but, I have a simple Set type:

```
 (define-type (Set :a)
    "A set of type :A elements.

     Sets are simple wrappers of lists, so they inherit all the performance
     limitations of lists. The maximum set size is system dependent."

    (Set (List :a)))
```

And I'd like to implement `Functor` for the type (just return a empty set for this example):

```
  (define-instance (Functor Set)

    (define (map fn set)
      "Map the function FN over SET."

      (empty-set)))
```

Coalton wants a constraint of `(Eq :b)` on the `map` function, which is fair enough. How do I manage type constraints in Coalton instances like this since `define-instance` doesn't support `declare` forms?

**Comment by eliaslfox** [comment, marked as answer]
A couple things:

Unfortunately there isn't a way to add a constraint where you want it.

The `empty-set` function probably doesn't need an `Eq :a` constraint.

I don't think sets have a valid functor instance. Map should obey `(map (compose f g))  ==  (compose (map f) (map g))`. Set's will remove elements if they compare equal. See [Functor](https://wiki.haskell.org/Functor#Functor_Laws) for more details.

**Reply by ornamentist** [reply]
Very true, I am pushing against the semantics of Functor, but I’m wondering if I’d have the same issue with say my own Vector or Matrix or Sequence types.

**Reply by eliaslfox** [reply]
Those other container types don't need extra constraints to map their elements.

**Reply by ornamentist** [reply]
I think I understand now why `Traversable` was invented (no help for Set but now I see why it's needed). Thanks again for your patient handling of Coalton discussions.
